File: images/fc_wrappers/fc_wrapper.js
// This is a wrapper which after bringing up a Node runtime
// signals Firecracker VMM to allow a snapshot to be taken
// if requested.
// Then it mounts application file system at /my-app,
// imports and executes application

const child_process = require('child_process');
// This requires ts.node be placed in the same directory as this wrapper
const ts = require('./ts.node');
const os = require('os');
const vcpu_count = os.cpus().length;
for (let i = 1; i < vcpu_count; i++) {
    child_process.exec(`taskset -c ${i} /srv/ts 124 1008`);
}
child_process.exec(`taskset -c 0 /srv/ts 124 1008`);

// mount app fs
child_process.execSync('mount /dev/vdb /my-app');
// This portio signals VMM to record boot up time
// if we are booting from a snapshot
ts.ts(126, 0x03f0);

app = require('/my-app/trigger.js')
app()

File: images/fc_wrappers/fc_wrapper.py
"""
This is a wrapper which after bringing up the python runtime
signals Firecracker VMM to allow a snapshot to be taken if
requested.
Then it mounts application file system at /my-app,
imports application as a module named trigger,
and executes application by calling its main()
"""
from importlib import import_module
import sys
import subprocess
from portio import outl
import os

# snapshot signal
for i in range(1, os.cpu_count()):
    subprocess.Popen('taskset -c %d /srv/ts 124 1008'%(i), shell=True)
subprocess.Popen('taskset -c 0 /srv/ts 124 1008', shell=True)

try:
    subprocess.check_call('mount /dev/vdb /my-app', shell=True)
except subprocess.CalledProcessError:
    outl(125, 0x03f0)
outl(126, 0x03f0)

sys.path.append('/my-app')
APP = import_module('trigger')
sys.path.pop()
APP.main()
